dataprochub/app.py

```diff
@@ -13,18 +13,35 @@
 # limitations under the License.
 
 """ Replaces default JupyterHub.app to handle redirects. """
-
+import re
 from jupyterhub.app import JupyterHub
 from jupyterhub.handlers.base import UserUrlHandler
+from jupyterhub.objects import Server
 
 
 class DataprocHubUserUrlHandler(UserUrlHandler):
   """ Extends UserUrlHandler to redirect user once spawn is done. """
 
   async def _redirect_to_user_server(self, user, spawner):
     self.statsd.incr('redirects.user_after_login')
-    redirect_url = user.spawners[spawner.name].component_gateway_url
-    self.log.debug(f'# spawner.name is `{spawner.name}`.')
+    redirect_url = None
+    tmp_spawner = user.spawners[spawner.name]
+    tmp_spawner_server = tmp_spawner._server
+    # When restarting the Docker container, spawner._server is a Server instead
+    # of being a DataprocHubServer. We extract the component_gateway_url from
+    # that Server(
+    #   url=https://<CG_URL>:443/user/matthieum/,
+    #   bind_url=https://<CG_URL>:443/user/matthieum/
+    # )
+    self.log.info(f'# spawner._server value is {tmp_spawner_server}')
+    if type(tmp_spawner_server) == Server:
```

Comment:
Does it make more sense to reverse this check? If type is DataprocHubServer, use the component_gateway_url, otherwise extract the URL.